cdap-security/src/main/java/io/cdap/cdap/security/authorization/InternalAccessEnforcer.java

```diff
@@ -0,0 +1,126 @@
+import io.cdap.cdap.proto.security.Permission;
+import io.cdap.cdap.proto.security.Principal;
+import io.cdap.cdap.security.auth.AccessToken;
+import io.cdap.cdap.security.auth.InvalidTokenException;
+import io.cdap.cdap.security.auth.TokenManager;
+import io.cdap.cdap.security.auth.UserIdentity;
+import io.cdap.cdap.security.auth.context.SystemAuthenticationContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.Set;
+
+public class InternalAccessEnforcer extends AbstractAccessEnforcer {
+  private static final Logger LOG = LoggerFactory.getLogger(InternalAccessEnforcer.class);
+
+  private final TokenManager tokenManager;
+  private final Codec<AccessToken> accessTokenCodec;
+
+  @Inject
+  InternalAccessEnforcer(CConfiguration cConf, TokenManager tokenManager, Codec<AccessToken> accessTokenCodec) {
+    super(cConf);
+    this.tokenManager = tokenManager;
+    this.accessTokenCodec = accessTokenCodec;
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    LOG.debug("Internal enforce check for entity {} for principal {} on permissions {}", entity, principal,
+              permissions);
+    validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+  }
+
+  @Override
+  public void enforceOnParent(EntityType entityType, EntityId parentId, Principal principal, Permission permission)
+    throws AccessException {
+    LOG.debug("Internal parent enforce check for entityType {} on parent entity {} for principal {} on permission {}",
+              entityType, parentId, principal, permission);
+    validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+  }
+
+  @Override
+  public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal)
+    throws AccessException {
+    LOG.debug("Internal isVisible check for entities {} for principal {}", entityIds, principal);
+    try {
+      validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+      return entityIds;
+    } catch (AccessException e) {
+      return Collections.emptySet();
+    }
+  }
+
+  private void validateAccessTokenAndIdentity(String principalName, Credential credential) throws AccessException {
+    if (credential == null) {
+      throw new IllegalStateException("Attempted to internally enforce access on null credential");
+    }
+    if (!credential.getType().equals(Credential.CredentialType.INTERNAL)) {
+      throw new IllegalStateException("Attempted to internally enforce access on non-internal credential type");
+    }
+    AccessToken accessToken;
+    try {
+      accessToken = accessTokenCodec.decode(Base64.getDecoder().decode(credential.getValue()));
+    } catch (IOException e) {
+      LOG.debug("Access token deserialization failure:", e);
```

Comment:
Don't log and throw. It will be creating duplicating logs. Only throw and let the higher level to log (and not rethrowing).

Reply:
Is it ok to log to error (or warn) if there are cases we don't expect to regularly happen?

Reply:
Fixed